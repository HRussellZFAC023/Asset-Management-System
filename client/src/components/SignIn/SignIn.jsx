import React from "react";
import "./SignIn.scss";
import Particles from "react-particles-js";
import particlesOptions from "../../consts/particles";
import { Link } from "react-router-dom";

class SignIn extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      signInEmail: '',
      signInPassword: '',
      errors: {
        email: '',
        password: '',
      },
      errorMessage: 'No error',
      showError: false
    }
  }

  handleChange = (event) => {
    this.setState({ showError: false })
    // eslint-disable-next-line
    const validEmailRegex = RegExp(/^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i);
    event.preventDefault();
    const { name, value } = event.target;
    let errors = this.state.errors;

    switch (name) {
      case 'email':
        errors.email =
          validEmailRegex.test(value)
            ? ''
            : 'Email is not valid!';
        break;
      case 'password':
        errors.password =
          value.length < 1
            ? 'Please insert your password!'
            : '';
        break;
      default:
        break;
    }
    this.setState({ errors, [name]: value });
  }

  validateForm = (errors) => {
    let valid = true;
    Object.values(errors).forEach(
      (val) => val.length > 0 && (valid = false)
    );
    return valid;
  }

  formIsValid = () => {
    const { signInEmail, signInPassword, errors } = this.state;

    if (signInEmail.length < 1) {
      this.setState(prevState =>
        ({ errors: { ...prevState.errors, email: 'Please insert your email' } }));
      return false;
    }
    if (signInPassword.length < 1) {
      this.setState(prevState =>
        ({ errors: { ...prevState.errors, password: 'Please insert your password' } }));
      return false;
    }

    if (signInPassword.length < 8) {
      this.setState(prevState =>
        ({ errors: { ...prevState.errors, password: 'Your password must be min 8 characters' } }));
      return false;
    }

    if (!errors.email.length && !errors.password.length) {
      this.clearErrors();
      return true;
    } else {
      return false;
    }
  }

  toggleError = () => {
    if (this.state.showError) {
      this.setState({ showError: false })
    } else {
      this.setState({ showError: true })
    }
  }

  displayError = (errorMessage) => {
    if (this.state.showError) {
      alert(errorMessage);
    }
  }

  clearErrors = () => {
    this.setState(prevState =>
      ({ errors: { ...prevState.errors, email: '' } }));
    this.setState(prevState =>
      ({ errors: { ...prevState.errors, password: '' } }));
  }

  handleOnEmailChange = (event) => {
    this.setState({ signInEmail: event.target.value })
    this.clearErrors();
    this.handleChange(event)
  }

  handleOnPasswordChange = (event) => {
    this.setState({ signInPassword: event.target.value })
    this.clearErrors()
    this.handleChange(event)
  }

  handleOnSubmitSignIn = () => {
    const { signInEmail, signInPassword } = this.state
    if (this.formIsValid()) {
      alert('Form submitted!');
      console.log('email: ' + signInEmail);
      console.log('password: ' + signInPassword);
    }
  }

  handleOnEnterKeyPress = (target) => {
    if (target.charCode === 13) {
      this.handleOnSubmitSignIn();
    }
  }

  render() {
    const { errors } = this.state;
    return (
      <div className="columns is-vcentered parent-form">
        <div className="login sidebar column is-one-third-widescreen is-one-quarter-fullhd">
          <section className="section is-centered">
            <div className="has-text-centered">
              <img
                onClick={() => window.location.replace("/")}
                className="login-logo"
                alt='logo' src="https://image.flaticon.com/icons/svg/615/615598.svg"
              />
            </div>
            <h1 className="title is-4">Sign In</h1>
            <div className="field">
              <label htmlFor="email" className="label">
                Email
              </label>
              <div className="control has-icons-left">
                <input
                  name="email"
                  type="email"
                  id="inputEmail"
                  placeholder="e.g. bobsmith@gmail.com"
                  className="input form-control"
                  onChange={this.handleOnEmailChange}
                  onKeyPress={this.handleOnEnterKeyPress}
                  required autoFocus
                />
                {errors.email.length > 0 && <span className='error'>{errors.email}</span>}
                <span className="icon is-small is-left">
                  <i className="fa fa-envelope" />
                </span>
              </div>
            </div>
            <div className="field">
              <label htmlFor="password" className="label">
                Password
              </label>
              <div className="control has-icons-left">
                <input
                  name="password"
                  type="password"
                  id="inputPassword"
                  placeholder="*******"
                  className="input form-control"
                  onChange={this.handleOnPasswordChange}
                  onKeyPress={this.handleOnEnterKeyPress}
                  required
                />
                {errors.password.length > 0 && <span className='mb-2 error'>{errors.password}</span>}
                <span className="icon is-small is-left">
                  <i className="fa fa-lock" />
                </span>
              </div>
            </div>
            <div className="field">
              <label htmlFor="checkbox" className="checkbox">
                <input type="checkbox" id='checkbox' />
                {" Remember me"}
              </label>
            </div>
            <div className="field">
              <button className="button button-login is-primary is-fullwidth" onClick={this.handleOnSubmitSignIn}>Sign In</button>
            </div>
            <div className='not-registered is-centered'>
              <p>Donâ€™t have an account?</p>
              <Link to='/signup'>Sign up</Link>
            </div>
          </section>
        </div>
        <div className="interactive-bg has-background-primary column">
          <Particles className='particles' params={particlesOptions} />
        </div>
      </div>
    );
  }
}

export default SignIn;
