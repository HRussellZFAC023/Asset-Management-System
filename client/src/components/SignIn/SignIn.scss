.parent-form {
  height: 100vh;
}

.login-box {
  margin-top: 7vh;
}

.not-registered {
  display: flex;
  a {
    padding-left: 5px;
  }
}

.error {
  display: flex;
  text-align: left;
  font-size: 0.7em;
  color: black;
}

.interactive-bg {
  height: 100%;
  background-image: 
    linear-gradient(transparent 11px, rgba(0, 0, 0, 0.10) 12px, transparent 12px),
    linear-gradient(90deg, transparent 11px, rgba(0, 0, 0, 0.10) 12px, transparent 12px);
  background-size: 100% 12px, 12px 100%;
  -webkit-box-shadow: inset 24px 4px 64px -24px rgba(71,71,71,1);
  -moz-box-shadow: inset 24px 4px 64px -24px rgba(71,71,71,1);
  box-shadow: inset 24px 4px 64px -24px rgba(71,71,71,1);
  padding: 0px;
}

.input {
  border-radius: 50px;
}

.home-link { 
  cursor: pointer;
}

.button-login {
  margin-top: 20px;
  margin-bottom: 20px;
  min-width: 150px;
}

.login-logo {
  margin: 0 auto;
  margin-bottom: 50px;
  max-height: 100px;
  cursor: pointer;
}

.columns{
  margin: 0px;
}

@media (min-width: 770px) {
  .sidebar {
    margin-top: -100px;
   }
  .interactive-bg {
    visibility: visible;
  }
}

@media (max-width: 769px) {
  .interactive-bg{
    visibility: hidden;
  }
  .interactive-bg {
    display: none;
  }
}

.particles {
  width: 100%;
  height: 100%;
}