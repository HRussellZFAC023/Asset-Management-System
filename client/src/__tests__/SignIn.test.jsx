import React from 'react';
import { shallow, mount } from 'enzyme';
import SignIn from '../components/sign-in/sign-in';

it('expect to render SignIn component', () => {
  expect(shallow(<SignIn />).length).toEqual(1)
})

it('expect html elements to render correctly', () => {
  const wrapper = mount(<SignIn />);
  expect(wrapper.html()).toMatchSnapshot();
})