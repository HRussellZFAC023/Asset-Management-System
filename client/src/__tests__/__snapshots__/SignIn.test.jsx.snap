// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect html elements to render correctly 1`] = `"<section class=\\"hero has-background-info my-fullheight square\\"><div class=\\"hero-body\\"><div class=\\"container\\"><div class=\\"columns is-centered\\"><div class=\\"column is-6-tablet is-5-desktop is-4-widescreen is-large login-box\\"><div class=\\"box\\"><h1 class=\\"title is-4\\">Sign In</h1><div class=\\"field\\"><label for=\\"email\\" class=\\"label\\">Email</label><div class=\\"control has-icons-left\\"><input name=\\"email\\" type=\\"email\\" id=\\"inputEmail\\" placeholder=\\"e.g. bobsmith@gmail.com\\" class=\\"input form-control\\" required=\\"\\"><span class=\\"icon is-small is-left\\"><i class=\\"fa fa-envelope\\"></i></span></div></div><div class=\\"field\\"><label for=\\"password\\" class=\\"label\\">Password</label><div class=\\"control has-icons-left\\"><input name=\\"password\\" type=\\"password\\" id=\\"inputPassword\\" placeholder=\\"*******\\" class=\\"input form-control\\" required=\\"\\"><span class=\\"icon is-small is-left\\"><i class=\\"fa fa-lock\\"></i></span></div></div><div class=\\"field\\"><label for=\\"checkbox\\" class=\\"checkbox\\"><input type=\\"checkbox\\" id=\\"checkbox\\"> Remember me</label></div><div class=\\"field\\"><button class=\\"button is-primary is-fullwidth\\">Sign In</button></div><div class=\\"not-registered is-centered\\"><p>Donâ€™t have an account?</p><a href=\\"/signup\\">Sign up</a></div></div></div></div></div></div></section>"`;
